// @astrojs/ts-plugin/dist/language-service/find-references.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  decorateFindReferences: () => decorateFindReferences
});
var import_utils = __toModule(require("../utils.js"));
function decorateFindReferences(ls, snapshotManager, logger) {
  decorateGetReferencesAtPosition(ls, snapshotManager, logger);
  _decorateFindReferences(ls, snapshotManager, logger);
}
function _decorateFindReferences(ls, snapshotManager, logger) {
  const findReferences = ls.findReferences;
  ls.findReferences = (fileName, position) => {
    const references = findReferences(fileName, position);
    return references == null ? void 0 : references.map((reference) => {
      const snapshot = snapshotManager.get(reference.definition.fileName);
      if (!(0, import_utils.isAstroFilePath)(reference.definition.fileName) || !snapshot) {
        return reference;
      }
      const textSpan = snapshot.getOriginalTextSpan(reference.definition.textSpan);
      if (!textSpan) {
        return null;
      }
      return {
        definition: __spreadProps(__spreadValues({}, reference.definition), {
          textSpan,
          originalTextSpan: void 0
        }),
        references: mapReferences(reference.references, snapshotManager, logger)
      };
    }).filter(import_utils.isNotNullOrUndefined);
  };
}
function decorateGetReferencesAtPosition(ls, snapshotManager, logger) {
  const getReferencesAtPosition = ls.getReferencesAtPosition;
  ls.getReferencesAtPosition = (fileName, position) => {
    const references = getReferencesAtPosition(fileName, position);
    return references && mapReferences(references, snapshotManager, logger);
  };
}
function mapReferences(references, snapshotManager, logger) {
  return references.map((reference) => {
    const snapshot = snapshotManager.get(reference.fileName);
    if (!(0, import_utils.isAstroFilePath)(reference.fileName) || !snapshot) {
      return reference;
    }
    const textSpan = snapshot.getOriginalTextSpan(reference.textSpan);
    if (!textSpan) {
      return null;
    }
    logger.debug("Find references; map textSpan: changed", reference.textSpan, "to", textSpan);
    return __spreadProps(__spreadValues({}, reference), {
      textSpan,
      contextSpan: void 0,
      originalTextSpan: void 0,
      originalContextSpan: void 0
    });
  }).filter(import_utils.isNotNullOrUndefined);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  decorateFindReferences
});
