// @astrojs/ts-plugin/dist/astro-snapshots.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};
__markAsModule(exports);
__export(exports, {
  AstroSnapshot: () => AstroSnapshot,
  AstroSnapshotManager: () => AstroSnapshotManager
});
var import_astro2tsx = __toModule(require("./astro2tsx.js"));
var import_source_mapper = __toModule(require("./source-mapper.js"));
var import_utils = __toModule(require("./utils.js"));
class AstroSnapshot {
  constructor(typescript, fileName, astroCode, mapper, logger, isTsFile) {
    this.typescript = typescript;
    this.fileName = fileName;
    this.astroCode = astroCode;
    this.mapper = mapper;
    this.logger = logger;
    this.isTsFile = isTsFile;
    this.convertInternalCodePositions = false;
  }
  update(astroCode, mapper) {
    this.astroCode = astroCode;
    this.mapper = mapper;
    this.lineOffsets = void 0;
    this.log("Updated Snapshot");
  }
  getOriginalTextSpan(textSpan) {
    if (!(0, import_utils.isNoTextSpanInGeneratedCode)(this.getText(), textSpan)) {
      return null;
    }
    const start = this.getOriginalOffset(textSpan.start);
    if (start === -1) {
      return null;
    }
    return {
      start,
      length: textSpan.length
    };
  }
  getOriginalOffset(generatedOffset) {
    if (!this.scriptInfo) {
      return generatedOffset;
    }
    this.toggleMappingMode(true);
    const lineOffset = this.scriptInfo.positionToLineOffset(generatedOffset);
    this.debug("try convert offset", generatedOffset, "/", lineOffset);
    const original = this.mapper.getOriginalPosition({
      line: lineOffset.line - 1,
      character: lineOffset.offset - 1
    });
    this.toggleMappingMode(false);
    if (original.line === -1) {
      return -1;
    }
    const originalOffset = this.scriptInfo.lineOffsetToPosition(original.line + 1, original.character + 1);
    this.debug("converted offset to", original, "/", originalOffset);
    return originalOffset;
  }
  setAndPatchScriptInfo(scriptInfo) {
    scriptInfo.scriptKind = this.typescript.ScriptKind.TSX;
    const positionToLineOffset = scriptInfo.positionToLineOffset.bind(scriptInfo);
    scriptInfo.positionToLineOffset = (position) => {
      if (this.convertInternalCodePositions) {
        const lineOffset2 = positionToLineOffset(position);
        this.debug("positionToLineOffset for generated code", position, lineOffset2);
        return lineOffset2;
      }
      const lineOffset = this.positionAt(position);
      this.debug("positionToLineOffset for original code", position, lineOffset);
      return {line: lineOffset.line + 1, offset: lineOffset.character + 1};
    };
    const lineOffsetToPosition = scriptInfo.lineOffsetToPosition.bind(scriptInfo);
    scriptInfo.lineOffsetToPosition = (line, offset) => {
      if (this.convertInternalCodePositions) {
        const position2 = lineOffsetToPosition(line, offset);
        this.debug("lineOffsetToPosition for generated code", {line, offset}, position2);
        return position2;
      }
      const position = this.offsetAt({line: line - 1, character: offset - 1});
      this.debug("lineOffsetToPosition for original code", {line, offset}, position);
      return position;
    };
    this.scriptInfo = scriptInfo;
    this.log("patched scriptInfo");
  }
  positionAt(offset) {
    offset = this.clamp(offset, 0, this.astroCode.length);
    const lineOffsets = this.getLineOffsets();
    let low = 0;
    let high = lineOffsets.length;
    if (high === 0) {
      return {line: 0, character: offset};
    }
    while (low < high) {
      const mid = Math.floor((low + high) / 2);
      if (lineOffsets[mid] > offset) {
        high = mid;
      } else {
        low = mid + 1;
      }
    }
    const line = low - 1;
    return {line, character: offset - lineOffsets[line]};
  }
  offsetAt(position) {
    const lineOffsets = this.getLineOffsets();
    if (position.line >= lineOffsets.length) {
      return this.astroCode.length;
    } else if (position.line < 0) {
      return 0;
    }
    const lineOffset = lineOffsets[position.line];
    const nextLineOffset = position.line + 1 < lineOffsets.length ? lineOffsets[position.line + 1] : this.astroCode.length;
    return this.clamp(nextLineOffset, lineOffset, lineOffset + position.character);
  }
  getLineOffsets() {
    if (this.lineOffsets) {
      return this.lineOffsets;
    }
    const lineOffsets = [];
    const text = this.astroCode;
    let isLineStart = true;
    for (let i = 0; i < text.length; i++) {
      if (isLineStart) {
        lineOffsets.push(i);
        isLineStart = false;
      }
      const ch = text.charAt(i);
      isLineStart = ch === "\r" || ch === "\n";
      if (ch === "\r" && i + 1 < text.length && text.charAt(i + 1) === "\n") {
        i++;
      }
    }
    if (isLineStart && text.length > 0) {
      lineOffsets.push(text.length);
    }
    this.lineOffsets = lineOffsets;
    return lineOffsets;
  }
  clamp(num, min, max) {
    return Math.max(min, Math.min(max, num));
  }
  log(...args) {
    this.logger.log("AstroSnapshot:", this.fileName, "-", ...args);
  }
  debug(...args) {
    this.logger.debug("AstroSnapshot:", this.fileName, "-", ...args);
  }
  toggleMappingMode(convertInternalCodePositions) {
    this.convertInternalCodePositions = convertInternalCodePositions;
  }
  getText() {
    var _a;
    const snapshot = (_a = this.scriptInfo) == null ? void 0 : _a.getSnapshot();
    if (!snapshot) {
      return "";
    }
    return snapshot.getText(0, snapshot.getLength());
  }
}
class AstroSnapshotManager {
  constructor(typescript, projectService, logger) {
    this.typescript = typescript;
    this.projectService = projectService;
    this.logger = logger;
    this.snapshots = new Map();
    this.patchProjectServiceReadFile();
  }
  get(fileName) {
    return this.snapshots.get(fileName);
  }
  create(fileName) {
    if (this.snapshots.has(fileName)) {
      return this.snapshots.get(fileName);
    }
    const scriptInfo = this.projectService.getOrCreateScriptInfoForNormalizedPath(this.typescript.server.toNormalizedPath(fileName), false);
    if (!scriptInfo) {
      this.logger.log("Was not able get snapshot for", fileName);
      return;
    }
    try {
      scriptInfo.getSnapshot();
    } catch (e) {
      this.logger.log("Loading Snapshot failed", fileName);
    }
    const snapshot = this.snapshots.get(fileName);
    if (!snapshot) {
      this.logger.log("Astro snapshot was not found after trying to load script snapshot for", fileName);
      return;
    }
    snapshot.setAndPatchScriptInfo(scriptInfo);
    this.snapshots.set(fileName, snapshot);
    return snapshot;
  }
  patchProjectServiceReadFile() {
    const readFile = this.projectService.host.readFile;
    this.projectService.host.readFile = (path) => {
      if ((0, import_utils.isAstroFilePath)(path)) {
        this.logger.debug("Read Astro file:", path);
        const astroCode = readFile(path) || "";
        try {
          const isTsFile = true;
          const result = (0, import_astro2tsx.astro2tsx)(astroCode, {
            filename: path.split("/").pop(),
            isTsFile
          });
          const existingSnapshot = this.snapshots.get(path);
          if (existingSnapshot) {
            existingSnapshot.update(astroCode, new import_source_mapper.SourceMapper(result.map.mappings));
          } else {
            this.snapshots.set(path, new AstroSnapshot(this.typescript, path, astroCode, new import_source_mapper.SourceMapper(result.map.mappings), this.logger, isTsFile));
          }
          this.logger.log("Successfully read Astro file contents of", path);
          return result.code;
        } catch (e) {
          this.logger.log("Error loading Astro file:", path);
          this.logger.debug("Error:", e);
        }
      } else {
        return readFile(path);
      }
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AstroSnapshot,
  AstroSnapshotManager
});
