// @astrojs/ts-plugin/dist/utils.js
var __defProp = Object.defineProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
__markAsModule(exports);
__export(exports, {
  ensureRealAstroFilePath: () => ensureRealAstroFilePath,
  isAstroFilePath: () => isAstroFilePath,
  isInGeneratedCode: () => isInGeneratedCode,
  isNoTextSpanInGeneratedCode: () => isNoTextSpanInGeneratedCode,
  isNotNullOrUndefined: () => isNotNullOrUndefined,
  isVirtualAstroFilePath: () => isVirtualAstroFilePath,
  replaceDeep: () => replaceDeep,
  toRealAstroFilePath: () => toRealAstroFilePath
});
function isAstroFilePath(filePath) {
  return filePath.endsWith(".astro");
}
function isVirtualAstroFilePath(filePath) {
  return filePath.endsWith(".astro.ts");
}
function toRealAstroFilePath(filePath) {
  return filePath.slice(0, -".ts".length);
}
function ensureRealAstroFilePath(filePath) {
  return isVirtualAstroFilePath(filePath) ? toRealAstroFilePath(filePath) : filePath;
}
function isNotNullOrUndefined(val) {
  return val !== void 0 && val !== null;
}
function isInGeneratedCode(text, start, end) {
  const lineStart = text.lastIndexOf("\n", start);
  const lineEnd = text.indexOf("\n", end);
  const lastStart = text.substring(lineStart, start).lastIndexOf("/*\u03A9ignore_start\u03A9*/");
  const lastEnd = text.substring(lineStart, start).lastIndexOf("/*\u03A9ignore_end\u03A9*/");
  return lastStart > lastEnd && text.substring(end, lineEnd).includes("/*\u03A9ignore_end\u03A9*/");
}
function isNoTextSpanInGeneratedCode(text, span) {
  return !isInGeneratedCode(text, span.start, span.start + span.length);
}
function replaceDeep(obj, searchStr, replacementStr) {
  return _replaceDeep(obj);
  function _replaceDeep(_obj) {
    if (typeof _obj === "string") {
      return _obj.replace(searchStr, replacementStr);
    }
    if (Array.isArray(_obj)) {
      return _obj.map((entry) => _replaceDeep(entry));
    }
    if (typeof _obj === "object") {
      return Object.keys(_obj).reduce((_o, key) => {
        _o[key] = _replaceDeep(_obj[key]);
        return _o;
      }, {});
    }
    return _obj;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ensureRealAstroFilePath,
  isAstroFilePath,
  isInGeneratedCode,
  isNoTextSpanInGeneratedCode,
  isNotNullOrUndefined,
  isVirtualAstroFilePath,
  replaceDeep,
  toRealAstroFilePath
});
